engine/api/workflow/dao_run.go

```diff
@@ -335,3 +336,50 @@ func nextRunNumber(db gorp.SqlExecutor, w *sdk.Workflow) (int64, error) {
 	log.Debug("nextRunNumber> %s/%s %d", w.ProjectKey, w.Name, i)
 	return int64(i), nil
 }
+
+// PurgeWorkflowRun mark all worfklow run to delete
+func PurgeWorkflowRun(db gorp.SqlExecutor, wf sdk.Workflow, branch string) error {
+	if wf.HistoryLength == 0 {
+		log.Warning("PurgeWorkflowRun> history length equals 0, skipping purge")
+		return nil
+	}
+
+	queryUpdate := `UPDATE workflow_run SET to_delete = true WHERE workflow_run.id IN ( %s )`
+	args := []interface{}{wf.ID}
+	if branch == "" {
+		queryUpdate = fmt.Sprintf(queryUpdate, `SELECT workflow_run.id FROM workflow_run
+			LEFT JOIN workflow_run_tag ON workflow_run.id = workflow_run_tag.workflow_run_id
+			WHERE workflow_run.workflow_id = $1
+			AND (workflow_run_tag.tag IS NULL
+				OR (workflow_run_tag.tag = 'git.branch' AND (workflow_run_tag.value IS NULL OR workflow_run_tag.value = ''))
+			) ORDER BY workflow_run.id DESC OFFSET $2 ROWS
+		`)
+	} else {
+		queryUpdate = fmt.Sprintf(queryUpdate, `SELECT workflow_run.id FROM workflow_run
+			JOIN workflow_run_tag ON workflow_run.id = workflow_run_tag.workflow_run_id
+			WHERE workflow_run.workflow_id = $1
+			AND workflow_run_tag.tag = 'git.branch'
+			AND workflow_run_tag.value = $2
+			ORDER BY workflow_run.id DESC
+			OFFSET $3 ROWS
+		`)
+		args = append(args, branch)
+	}
+	args = append(args, wf.HistoryLength)
+
+	if _, err := db.Exec(queryUpdate, args...); err != nil {
+		log.Warning("PurgeWorkflowRun> Unable to update workflow run for purge for id %d, branch %s and history length %d : %s", wf.ID, branch, wf.HistoryLength, err)
+	}
+	return nil
+}
+
+// DeleteWorkflowRunsHistory is useful to delete all the workflow run marked with to delete flag in db
+func DeleteWorkflowRunsHistory(db gorp.SqlExecutor) error {
+	query := `DELETE FROM workflow_run WHERE to_delete = true`
```

Comment:
You should select workflow_run_id then iterate over the result set and delete workflow run one by one: one transaction by deletion, because if one deletion failed, we still have to (try to) delete the other one